doc/src/gotchas.rst

```diff
@@ -758,6 +758,103 @@ that you want, and they will all be evaluated according to the function.
 
     See the Python docs for more information on function parameters.
 
+
+Classification of the objects
+=============================
+
+There are several ways of how SymPy object is classified.
+
+class
+-----
+
+Like any object, expression is an instance of class. You can get the class of
+the object with built-in `type()` function, and check it with `isinstance()` function.
+
+    >>> type(x + y)
+    <class 'sympy.core.add.Add'>
+    >>> isinstance(x + y, Add)
+    True
+
+Classes represent only the programmatic structures of the objects, and does not
+distinguish the mathematical difference between them. For example, the integral
+of number and the integral of matrix both have the class `Integral`, although the
+former is number and the latter is matrix.
+
+    >>> A = MatrixSymbol('A', 2, 2)
+    >>> type(Integral(1, x))
+    <class 'sympy.integrals.integrals.Integral'>
+    >>> type(Integral(A, x))
+    <class 'sympy.integrals.integrals.Integral'>
+
+typeclass
+---------
+
+Typeclass [1]_ indicates what mathematical object (in broad sense) does the
+expression represent. You can retrieve the typeclass of expression with
+`sympy.core.typeclass()` function.
+
+    >>> from sympy.core import typeclass
```

Comment:
Having spent the last few months working with "typeclasses" in lean, where they resemble haskell - I don't think this is the correct meaning of the word, and wikipedia confirms this with the phrasing "belongs to a typeclass".

For example, a number type would likely belong to all of `Ordered`, `HasAdd`, and  `Hashable` typeclasses.
Your implementation has each type assign itself a _single_ `typeclass`. This is not how typeclasses work at all - instead, they consist of external bits of code registering a variety of typeclasses (aking to multiple dispatch) for a variety of types.

[PEP544](https://www.python.org/dev/peps/pep-0544/) looks to me like a reasonable approximation of typeclasses for static-typing of typeclasses, while duck-typing is already the equivalent for implementing typeclasses.

I think you need to come up with a different name.

Reply:
@oscarbenjamin Any idea?

Reply:
I guess any name we go with would clash with something.

All the names have already been taken though: category, class, type, etc. Presumably that's why other people already came up with "typeclass".

Okay so synonyms from Macmillan:
```
Categories and types
SYNONYMS
type NOUN
a group of people or things with similar qualities or features that make them different from other groups

category NOUN
a group of people or things that have similar qualities

kind NOUN
a type of person or thing

sort NOUN
a group of people or things with the same qualities or features

variety NOUN
a type of thing

classification NOUN
a group into which someone or something is put because of the features that they have

grouping NOUN
a set of people or things that are considered as a group

taxonomy NOUN
a system for organizing similar things into groups

class NOUN
a group of things, animals, or people with similar features or qualities

cross section NOUN
a group that contains an example of all or most types of people or things
```
https://www.macmillandictionary.com/thesaurus-category/british/categories-and-types

From the definitions given it would naturally seem that "type" is most suited but it already has a meaning in Python where both `type` and `Type` are already used. It also most closely matches e.g. type theory:
https://en.wikipedia.org/wiki/Type_theory
Maybe referring to something like type theory is a mistake though. The differences might be more distracting than the similarities (as for typeclasses).

Synonyms of "type"  are "kind" and "variety". Since "variety" is already used in maths we could consider "kind". Then we have "NumberKind", "MatrixKind" and
```python
def sin(expr):
    """sine function

    parameters
    ==========

    expr: NumberKind
        Any complex number
    """
    if kind(expr) is not NumberKind:
        raise TypeError
```
It seems

Reply:
I changed the naming to `kind`.